In [35]:
import pandas as pd

In [36]:
data = pd.read_csv(r"/content/spambase.data")
data.head()

0  0.64  0.64.1  0.1  0.32   0.2   0.3   0.4   0.5   0.6  ...  0.40  \
0  0.21  0.28    0.50  0.0  0.14  0.28  0.21  0.07  0.00  0.94  ...  0.00   
1  0.06  0.00    0.71  0.0  1.23  0.19  0.19  0.12  0.64  0.25  ...  0.01   
2  0.00  0.00    0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.00   
3  0.00  0.00    0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.00   
4  0.00  0.00    0.00  0.0  1.85  0.00  0.00  1.85  0.00  0.00  ...  0.00   

    0.41  0.42  0.778   0.43   0.44  3.756   61   278  1  
0  0.132   0.0  0.372  0.180  0.048  5.114  101  1028  1  
1  0.143   0.0  0.276  0.184  0.010  9.821  485  2259  1  
2  0.137   0.0  0.137  0.000  0.000  3.537   40   191  1  
3  0.135   0.0  0.135  0.000  0.000  3.537   40   191  1  
4  0.223   0.0  0.000  0.000  0.000  3.000   15    54  1  

[5 rows x 58 columns]

In [37]:
data.shape

(4600, 58)

In [38]:
data.describe()

0         0.64       0.64.1          0.1         0.32  \
count  4600.000000  4600.000000  4600.000000  4600.000000  4600.000000   
mean      0.104576     0.212922     0.280578     0.065439     0.312222   
std       0.305387     1.290700     0.504170     1.395303     0.672586   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.420000     0.000000     0.382500   
max       4.540000    14.280000     5.100000    42.810000    10.000000   

               0.2          0.3          0.4          0.5          0.6  ...  \
count  4600.000000  4600.000000  4600.000000  4600.000000  4600.000000  ...   
mean      0.095922     0.114233     0.105317     0.090087     0.239465  ...   
std       0.273850     0.391480     0.401112     0.278643     0.644816  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
50%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
75%       0.000000     0.000000     0.000000     0.000000     0.160000  ...   
max       5.880000     7.270000    11.110000     5.260000    18.180000  ...   

              0.40         0.41         0.42        0.778         0.43  \
count  4600.000000  4600.000000  4600.000000  4600.000000  4600.000000   
mean      0.038583     0.139061     0.016980     0.268960     0.075827   
std       0.243497     0.270377     0.109406     0.815726     0.245906   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.065000     0.000000     0.000000     0.000000   
75%       0.000000     0.188000     0.000000     0.314250     0.052000   
max       4.385000     9.752000     4.081000    32.478000     6.003000   

              0.44        3.756           61           278            1  
count  4600.000000  4600.000000  4600.000000   4600.000000  4600.000000  
mean      0.044248     5.191827    52.170870    283.290435     0.393913  
std       0.429388    31.732891   194.912453    606.413764     0.488669  
min       0.000000     1.000000     1.000000      1.000000     0.000000  
25%       0.000000     1.588000     6.000000     35.000000     0.000000  
50%       0.000000     2.275500    15.000000     95.000000     0.000000  
75%       0.000000     3.705250    43.000000    265.250000     1.000000  
max      19.829000  1102.500000  9989.000000  15841.000000     1.000000  

[8 rows x 58 columns]

In [39]:
data['1'].value_counts()

0    2788
1    1812
Name: 1, dtype: int64

In [40]:
data.isnull().sum()

0         0
0.64      0
0.64.1    0
0.1       0
0.32      0
0.2       0
0.3       0
0.4       0
0.5       0
0.6       0
0.7       0
0.64.2    0
0.8       0
0.9       0
0.10      0
0.32.1    0
0.11      0
1.29      0
1.93      0
0.12      0
0.96      0
0.13      0
0.14      0
0.15      0
0.16      0
0.17      0
0.18      0
0.19      0
0.20      0
0.21      0
0.22      0
0.23      0
0.24      0
0.25      0
0.26      0
0.27      0
0.28      0
0.29      0
0.30      0
0.31      0
0.32.2    0
0.33      0
0.34      0
0.35      0
0.36      0
0.37      0
0.38      0
0.39      0
0.40      0
0.41      0
0.42      0
0.778     0
0.43      0
0.44      0
3.756     0
61        0
278       0
1         0
dtype: int64

In [41]:
from numpy import mean
from numpy import std
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.linear_model import Perceptron
from sklearn.model_selection import GridSearchCV

In [42]:
#splitting feature and target columns
X = data.drop('1',axis=1)
y= data['1']

In [43]:
#model defining
model = Perceptron()

In [44]:
#model evaluation
cro_val = RepeatedStratifiedKFold(n_splits =10, n_repeats = 5, random_state = 42)

In [45]:
score = cross_val_score(model, X, y, scoring='accuracy', cv = cro_val, n_jobs=1)
print('Mean Accuracy:')
print( "%.3f, (%.3f)" % (mean(score), std(score)) )

Mean Accuracy:
0.623, (0.102)


In [46]:
#hyperparameter tuning for learning rate
grid = dict()
grid['eta0'] = [0.0001, 0.001, 0.01, 0.1, 1.0]

In [47]:
search = GridSearchCV(model, grid, scoring='accuracy',cv= cro_val, n_jobs=-1)
results = search.fit(X,y)

In [48]:
print('Mean Accuracy: %.3f' % results.best_score_)
print('Config: %s' % results.best_params_)
means = results.cv_results_['mean_test_score']
para = results.cv_results_['params']
for mean, p in zip(means,para):
  print("--> %.3f with: %r" % (mean,p))

Mean Accuracy: 0.623
Config: {'eta0': 0.0001}
--> 0.623 with: {'eta0': 0.0001}
--> 0.623 with: {'eta0': 0.001}
--> 0.623 with: {'eta0': 0.01}
--> 0.623 with: {'eta0': 0.1}
--> 0.623 with: {'eta0': 1.0}


In [49]:
#hyperparameter tuning for epochs 
grid = dict()
grid['max_iter'] = [1,10,100,1000,10000]

In [50]:
# perceptron model with best learning rate
model = Perceptron(eta0 = 0.0001)

In [51]:
search = GridSearchCV(model, grid, scoring='accuracy', cv=cro_val, n_jobs=-1)
result = search.fit(X, y)
print('Mean Accuracy: %.3f' % result.best_score_)
print('Config: %s' % result.best_params_)

Mean Accuracy: 0.629
Config: {'max_iter': 10}


In [52]:
means = result.cv_results_['mean_test_score']
params = result.cv_results_['params']
for mean, param in zip(means, params):
    print(">%.3f with: %r" % (mean, param))

>0.569 with: {'max_iter': 1}
>0.629 with: {'max_iter': 10}
>0.623 with: {'max_iter': 100}
>0.623 with: {'max_iter': 1000}
>0.623 with: {'max_iter': 10000}


Conclusion:
Model with learning rate 0.0001 and epoch 10 gave the best result with mean accuracy 62.9%